File: src/symlink.ts
import { lstatSync, mkdirSync, readFileSync, readlinkSync, rmSync, symlinkSync } from "node:fs";
import type { FileSpec, ProviderConfig, ProvidersConfig } from "./types";
import { discoverAll } from "./discovery";
import { pathBasename, pathDirname, pathJoin, pathRelative } from "./paths";

export function syncSymlinks(
  sources: string[],
  selectedProviders: string[],
  cfg: ProvidersConfig,
  specSelector: (provider: ProviderConfig) => FileSpec | undefined,
  dryRun: boolean,
  verbose: boolean,
): { created: number; skipped: number; operations: string[] } {
  let created = 0;
  let skipped = 0;
  const operations: string[] = [];

  for (const sourcePath of sources) {
    const dir = pathDirname(sourcePath);
    const filename = pathBasename(sourcePath);

    for (const providerName of selectedProviders) {
      const provider = cfg.providers[providerName];
      const spec = specSelector(provider);
      if (!spec) {
        continue;
      }

      let result: { created: number; skipped: number; op?: string };
      if (!spec.dir) {
        result = createSymlink(dir, filename, spec.file, dryRun, verbose);
      } else {
        result = createSymlinkInDir(dir, filename, spec.dir, spec.file, dryRun, verbose);
      }
      created += result.created;
      skipped += result.skipped;
      if (result.op) {
        operations.push(result.op);
      }
    }
  }

  return { created, skipped, operations };
}

export function deleteManagedFiles(
  selectedProviders: string[],
  cfg: ProvidersConfig,
  sourceName: string,
  specSelector: (provider: ProviderConfig) => FileSpec | undefined,
  dryRun: boolean,
  verbose: boolean,
) {
  let deleted = 0;
  let notFound = 0;
  const operations: string[] = [];

  const allFiles = discoverAll(cfg, sourceName, specSelector);

  for (const file of allFiles) {
    const shouldDelete = selectedProviders.some((provider) => file.agent.toLowerCase() === provider.toLowerCase());
    if (!shouldDelete) {
      continue;
    }

    if (dryRun) {
      deleted += 1;
      if (verbose) {
        operations.push(`would delete: ${file.path}`);
      }
      continue;
    }

    try {
      rmSync(file.path);
      deleted += 1;
      if (verbose) {
        operations.push(`deleted: ${file.path}`);
      }
    } catch (error) {
      notFound += 1;
      if (verbose) {
        operations.push(`error: ${file.path} (${error})`);
      }
    }
  }

  formatRmSummary(deleted, notFound, verbose, operations);
}

function createSymlink(dir: string, source: string, target: string, dryRun: boolean, verbose: boolean) {
  const sourcePath = pathJoin(dir, source);
  const targetPath = pathJoin(dir, target);

  if (exists(targetPath)) {
    const info = lstatSync(targetPath);
    if (shouldSkipOrOverwrite(targetPath, source, info.isSymbolicLink(), sourcePath, dryRun)) {
      return { created: 0, skipped: 1, op: verbose ? `skipped: ${targetPath} (already correct)` : undefined };
    }
    if (!dryRun) {
      rmSync(targetPath);
    }
  }

  if (dryRun) {
    return { created: 1, skipped: 0, op: verbose ? `would create: ${targetPath} -> ${source}` : undefined };
  }

  try {
    symlinkSync(source, targetPath);
    return { created: 1, skipped: 0, op: verbose ? `created: ${targetPath} -> ${source}` : undefined };
  } catch (error) {
    return { created: 0, skipped: 1, op: verbose ? `error: ${targetPath} (${error})` : undefined };
  }
}

function createSymlinkInDir(
  dir: string,
  source: string,
  subdir: string,
  target: string,
  dryRun: boolean,
  verbose: boolean,
) {
  const subdirPath = pathJoin(dir, subdir);
  if (!exists(subdirPath)) {
    if (!dryRun) {
      mkdirSync(subdirPath, { recursive: true });
    }
  }

  const sourcePath = pathJoin(dir, source);
  const targetPath = pathJoin(subdirPath, target);
  const symTarget = pathRelative(subdirPath, sourcePath);

  if (exists(targetPath)) {
    const info = lstatSync(targetPath);
    if (shouldSkipOrOverwrite(targetPath, symTarget, info.isSymbolicLink(), sourcePath, dryRun)) {
      return { created: 0, skipped: 1, op: verbose ? `skipped: ${targetPath} (already correct)` : undefined };
    }
    if (!dryRun) {
      rmSync(targetPath);
    }
  }

  if (dryRun) {
    return { created: 1, skipped: 0, op: verbose ? `would create: ${targetPath} -> ${symTarget}` : undefined };
  }

  try {
    symlinkSync(symTarget, targetPath);
    return { created: 1, skipped: 0, op: verbose ? `created: ${targetPath} -> ${symTarget}` : undefined };
  } catch (error) {
    return { created: 0, skipped: 1, op: verbose ? `error: ${targetPath} (${error})` : undefined };
  }
}

function shouldSkipOrOverwrite(
  targetPath: string,
  expectedTarget: string,
  isSymlink: boolean,
  sourcePath: string,
  dryRun: boolean,
) {
  if (!exists(targetPath)) {
    return false;
  }

  if (isSymlink) {
    try {
      const link = readlinkSync(targetPath);
      if (link === expectedTarget) {
        return true;
      }
    } catch {
      return false;
    }
    if (!dryRun && !askForConfirmation(targetPath, expectedTarget)) {
      return true;
    }
    return false;
  }

  const sourceContent = safeReadFile(sourcePath);
  const targetContent = safeReadFile(targetPath);
  if (sourceContent === null || targetContent === null) {
    if (!dryRun && !askForConfirmation(targetPath, "source")) {
      return true;
    }
    return false;
  }

  if (sourceContent === targetContent) {
    return true;
  }

  if (!dryRun && !askForConfirmation(targetPath, "different version")) {
    return true;
  }
  return false;
}

function askForConfirmation(targetPath: string, reason: string): boolean {
  try {
    process.stdout.write(`\nFile already exists: ${targetPath} (${reason})\nOverwrite? (y/n): `);
  } catch {
    return false;
  }
  const response = readLine();
  if (!response) {
    return false;
  }
  const normalized = response.trim().toLowerCase();
  return normalized === "y" || normalized === "yes";
}

function readLine(): string | null {
  try {
    const data = readFileSync(0, "utf-8");
    if (!data) {
      return null;
    }
    const line = data.split(/\r?\n/)[0];
    return line === undefined ? null : line;
  } catch {
    return null;
  }
}

function safeReadFile(path: string): string | null {
  try {
    return readFileSync(path, "utf-8");
  } catch {
    return null;
  }
}

function exists(path: string): boolean {
  try {
    lstatSync(path);
    return true;
  } catch {
    return false;
  }
}

function formatRmSummary(deleted: number, notFound: number, verbose: boolean, operations: string[]) {
  console.log(`Files deleted: ${deleted}`);
  if (notFound > 0) {
    console.log(`Errors: ${notFound}`);
  }

  if (verbose && operations.length > 0) {
    console.log("Operations:");
    for (const op of operations) {
      console.log(op);
    }
  }
}

File: src/symlink.go
package main

import (
	"bufio"
	"bytes"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

// SyncResult holds the result of a sync operation
type SyncResult struct {
	Created    int
	Skipped    int
	Operations []string
}

func syncSymlinks(
	sources []string,
	selectedTools []string,
	cfg *ToolsConfig,
	specSelector func(ToolConfig) *FileSpec,
	dryRun bool,
	verbose bool,
) SyncResult {
	result := SyncResult{
		Created:    0,
		Skipped:    0,
		Operations: []string{},
	}

	for _, sourcePath := range sources {
		dir := pathDirname(sourcePath)
		filename := pathBasename(sourcePath)

		for _, toolName := range selectedTools {
			tool, ok := cfg.Tools[toolName]
			if !ok {
				continue
			}

			spec := specSelector(tool)
			if spec == nil {
				continue
			}

			var opResult SyncResult
			// Determine target filename:
			// - For same-directory symlinks (spec.Dir == ""), use spec.File
			// - For subdirectory symlinks (spec.Dir != ""), preserve source filename
			targetFile := filename
			if spec.Dir == "" {
				targetFile = spec.File
			}

			if spec.Dir == "" {
				opResult = createSymlink(dir, filename, targetFile, dryRun, verbose)
			} else {
				opResult = createSymlinkInDir(dir, filename, spec.Dir, targetFile, dryRun, verbose)
			}

			result.Created += opResult.Created
			result.Skipped += opResult.Skipped
			result.Operations = append(result.Operations, opResult.Operations...)
		}
	}

	return result
}

func deleteManagedFiles(
	selectedTools []string,
	cfg *ToolsConfig,
	sourceName string,
	specSelector func(ToolConfig) *FileSpec,
	dryRun bool,
	verbose bool,
) {
	deleted := 0
	notFound := 0
	operations := []string{}

	allFiles := discoverAll(cfg, sourceName, specSelector)

	for _, file := range allFiles {
		shouldDelete := false
		for _, tool := range selectedTools {
			if strings.EqualFold(file.Tool, tool) {
				shouldDelete = true
				break
			}
		}

		if !shouldDelete {
			continue
		}

		if dryRun {
			deleted++
			if verbose {
				operations = append(operations, fmt.Sprintf("would delete: %s", file.Path))
			}
			continue
		}

		err := os.Remove(file.Path)
		if err != nil {
			notFound++
			if verbose {
				operations = append(operations, fmt.Sprintf("error: %s (%v)", file.Path, err))
			}
		} else {
			deleted++
			if verbose {
				operations = append(operations, fmt.Sprintf("deleted: %s", file.Path))
			}
		}
	}

	formatRmSummary(deleted, notFound, verbose, operations)
}

// createSymlink creates a symlink in the same directory
func createSymlink(dir, source, target string, dryRun, verbose bool) SyncResult {
	result := SyncResult{}
	sourcePath := pathJoin(dir, source)
	targetPath := pathJoin(dir, target)

	if exists(targetPath) {
		if shouldSkipOrOverwrite(targetPath, source, sourcePath, dryRun) {
			result.Skipped = 1
			if verbose {
				result.Operations = append(result.Operations, fmt.Sprintf("skipped: %s (already correct)", targetPath))
			}
			return result
		}
		if !dryRun {
			os.Remove(targetPath)
		}
	}

	if dryRun {
		result.Created = 1
		if verbose {
			result.Operations = append(result.Operations, fmt.Sprintf("would create: %s -> %s", targetPath, source))
		}
		return result
	}

	err := os.Symlink(source, targetPath)
	if err != nil {
		result.Skipped = 1
		if verbose {
			result.Operations = append(result.Operations, fmt.Sprintf("error: %s (%v)", targetPath, err))
		}
	} else {
		result.Created = 1
		if verbose {
			result.Operations = append(result.Operations, fmt.Sprintf("created: %s -> %s", targetPath, source))
		}
	}

	return result
}

// createSymlinkInDir creates a symlink in a subdirectory
func createSymlinkInDir(dir, source, subdir, target string, dryRun, verbose bool) SyncResult {
	result := SyncResult{}

	// Get current working directory
	cwd, _ := os.Getwd()

	// Make paths absolute for comparison
	absSourceDir, _ := filepath.Abs(dir)
	absCwd, _ := filepath.Abs(cwd)

	// Determine project root - if source is deeper than CWD, use CWD
	projectRoot := absSourceDir
	if len(absSourceDir) > len(absCwd) && filepath.HasPrefix(absSourceDir, absCwd) {
		projectRoot = absCwd
	}

	// Create target directory relative to project root
	subdirPath := filepath.Join(projectRoot, subdir)

	if !exists(subdirPath) && !dryRun {
		os.MkdirAll(subdirPath, 0755)
	}

	// Get absolute source path
	sourcePath := filepath.Join(absSourceDir, source)
	targetPath := filepath.Join(subdirPath, target)
	symTarget, _ := filepath.Rel(subdirPath, sourcePath)

	if exists(targetPath) {
		if shouldSkipOrOverwrite(targetPath, symTarget, sourcePath, dryRun) {
			result.Skipped = 1
			if verbose {
				result.Operations = append(result.Operations, fmt.Sprintf("skipped: %s (already correct)", targetPath))
			}
			return result
		}
		if !dryRun {
			os.Remove(targetPath)
		}
	}

	if dryRun {
		result.Created = 1
		if verbose {
			result.Operations = append(result.Operations, fmt.Sprintf("would create: %s -> %s", targetPath, symTarget))
		}
		return result
	}

	err := os.Symlink(symTarget, targetPath)
	if err != nil {
		result.Skipped = 1
		if verbose {
			result.Operations = append(result.Operations, fmt.Sprintf("error: %s (%v)", targetPath, err))
		}
	} else {
		result.Created = 1
		if verbose {
			result.Operations = append(result.Operations, fmt.Sprintf("created: %s -> %s", targetPath, symTarget))
		}
	}

	return result
}

// shouldSkipOrOverwrite checks if a symlink should be skipped or overwritten
func shouldSkipOrOverwrite(targetPath, expectedTarget, sourcePath string, dryRun bool) bool {
	if !exists(targetPath) {
		return false
	}

	// Check if it's a symlink
	if isSymlink(targetPath) {
		link, err := os.Readlink(targetPath)
		if err == nil && link == expectedTarget {
			return true
		}

		if !dryRun && !askForConfirmation(targetPath, expectedTarget) {
			return true
		}
		return false
	}

	// Check if content matches
	sourceContent := safeReadFile(sourcePath)
	targetContent := safeReadFile(targetPath)

	if sourceContent == nil || targetContent == nil {
		if !dryRun && !askForConfirmation(targetPath, "source") {
			return true
		}
		return false
	}

	if bytes.Equal(sourceContent, targetContent) {
		return true
	}

	if !dryRun && !askForConfirmation(targetPath, "different version") {
		return true
	}
	return false
}

// askForConfirmation prompts the user for confirmation
func askForConfirmation(targetPath, reason string) bool {
	fmt.Printf("\nFile already exists: %s (%s)\nOverwrite? (y/n): ", targetPath, reason)

	reader := bufio.NewReader(os.Stdin)
	response, err := reader.ReadString('\n')
	if err != nil {
		return false
	}

	normalized := strings.ToLower(strings.TrimSpace(response))
	return normalized == "y" || normalized == "yes"
}

// safeReadFile safely reads a file
func safeReadFile(path string) []byte {
	content, err := os.ReadFile(path)
	if err != nil {
		return nil
	}
	return content
}

// exists checks if a path exists
func exists(path string) bool {
	_, err := os.Lstat(path)
	return err == nil
}

// formatRmSummary formats the removal summary
func formatRmSummary(deleted, notFound int, verbose bool, operations []string) {
	fmt.Printf("Files deleted: %d\n", deleted)
	if notFound > 0 {
		fmt.Printf("Errors: %d\n", notFound)
	}

	if verbose && len(operations) > 0 {
		fmt.Println("Operations:")
		for _, op := range operations {
			fmt.Println(op)
		}
	}
}
